### PR TITLE
Adopt the Vanniktech publish plugin

### PR DESCRIPTION
**Pull request by swankjesse**
It's really nice.

### CODE REVIEW

File: wire-library/wire-java-generator/build.gradle.kts [hunk 1]
```diff
@@ -24,3 +29,9 @@ dependencies {
   testImplementation(deps.assertj)
   testImplementation(deps.jimfs)
 }
+
+configure<MavenPublishBaseExtension> {
```
Comment (Egorand) [inline wire-library/wire-java-generator/build.gradle.kts:33]: Can maybe move this to the top level build file later? Can check which Kotlin plugin the subproject applies to pick the right `Kotlin*` config.

**Review by Egorand — APPROVED** [review]

File: wire-library/wire-gson-support/gradle.properties [hunk 2]
```diff
@@ -1,3 +0,0 @@
-POM_ARTIFACT_ID=wire-gson-support
-POM_NAME=Wire
```
Comment (jrodbx) [inline wire-library/wire-gson-support/gradle.properties:2]: Out of curiosity, why not keep the properties files and avoid the explicit pom configuration in the base plugin dsl?

**Comment by swankjesse** [comment]
```
> Task :wire-grpc-client:generateMetadataFileForKotlinMultiplatformPublication FAILED
Caching disabled for task ':wire-grpc-client:generateMetadataFileForKotlinMultiplatformPublication' because:
  Build cache is disabled
Task ':wire-grpc-client:generateMetadataFileForKotlinMultiplatformPublication' is not up-to-date because:
  Task.upToDateWhen is false.
:wire-grpc-client:generateMetadataFileForKotlinMultiplatformPublication (Thread[Execution worker for ':' Thread 3,5,main]) completed. Took 0.001 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':wire-grpc-client:generateMetadataFileForKotlinMultiplatformPublication'.
> Cannot publish module metadata where component artifacts are modified.

* Try:
Run with --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':wire-grpc-client:generateMetadataFileForKotlinMultiplatformPublication'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:200)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:198)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:179)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: org.gradle.api.artifacts.PublishException: Cannot publish module metadata where component artifacts are modified.
        at org.gradle.api.publish.maven.internal.publication.DefaultMavenPublication.checkThatArtifactIsPublishedUnmodified(DefaultMavenPublication.java:811)
        at org.gradle.api.publish.maven.internal.publication.DefaultMavenPublication.getPublishedFile(DefaultMavenPublication.java:768)
        at org.gradle.api.publish.internal.metadata.ModuleMetadataSpecBuilder.artifactFor(ModuleMetadataSpecBuilder.java:173)
        at org.gradle.api.publish.internal.metadata.ModuleMetadataSpecBuilder.artifactsOf(ModuleMetadataSpecBuilder.java:160)
        at org.gradle.api.publish.internal.metadata.ModuleMetadataSpecBuilder.variants(ModuleMetadataSpecBuilder.java:128)
        at org.gradle.api.publish.internal.metadata.ModuleMetadataSpecBuilder.build(ModuleMetadataSpecBuilder.java:92)
        at org.gradle.api.publish.internal.metadata.GradleModuleMetadataWriter.moduleMetadataSpecFor(GradleModuleMetadataWriter.java:84)
        at org.gradle.api.publish.tasks.GenerateModuleMetadata.computeModuleMetadataSpec(GenerateModuleMetadata.java:227)
        at org.gradle.internal.Try.ofFailable(Try.java:39)
        at org.gradle.api.publish.tasks.GenerateModuleMetadata.moduleMetadataSpec(GenerateModuleMetadata.java:223)
        at org.gradle.api.publish.tasks.GenerateModuleMetadata.computeInputState(GenerateModuleMetadata.java:219)
        at org.gradle.internal.Try.ofFailable(Try.java:39)
        at org.gradle.internal.serialization.Cached$Deferred.result(Cached.java:53)
        at org.gradle.internal.serialization.Cached$Deferred.get(Cached.java:48)
        at org.gradle.api.publish.tasks.GenerateModuleMetadata.inputState(GenerateModuleMetadata.java:291)
        at org.gradle.api.publish.tasks.GenerateModuleMetadata.hasAttachedComponent(GenerateModuleMetadata.java:205)
        at org.gradle.api.publish.tasks.GenerateModuleMetadata.lambda$new$b6ae1a43$1(GenerateModuleMetadata.java:94)
        at org.gradle.api.specs.AndSpec.isSatisfiedBy(AndSpec.java:50)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:44)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
```

File: wire-library/wire-java-generator/build.gradle.kts (hunk 1 above)
Comment (swankjesse) [reply wire-library/wire-java-generator/build.gradle.kts:33]: Yeah, I don’t mind this that much!

File: wire-library/wire-gson-support/gradle.properties (hunk 2 above)
Comment (swankjesse) [reply wire-library/wire-gson-support/gradle.properties:2]: I personally prefer the safety of the base DSL. IntelliJ automatically completes the available fields vs. properties files where I’m always anxious I’ve got a silent typo.

**Review by JakeWharton — APPROVED** [review]

**Pull request merged**